core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java

```diff
@@ -0,0 +1,85 @@
+package io.substrait.extendedexpression;
+
+import io.substrait.expression.Expression;
+import io.substrait.expression.proto.ProtoExpressionConverter;
+import io.substrait.extension.ExtensionCollector;
+import io.substrait.extension.ExtensionLookup;
+import io.substrait.extension.ImmutableExtensionLookup;
+import io.substrait.extension.ImmutableSimpleExtension;
+import io.substrait.extension.SimpleExtension;
+import io.substrait.proto.AggregateFunction;
+import io.substrait.proto.ExpressionReference;
+import io.substrait.proto.NamedStruct;
+import io.substrait.type.proto.ProtoTypeConverter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/** Converts from {@link io.substrait.proto.ExtendedExpression} to {@link ExtendedExpression} */
+public class ProtoExtendedExpressionConverter {
+  private final SimpleExtension.ExtensionCollection extensionCollection;
+
+  public ProtoExtendedExpressionConverter() throws IOException {
+    this(SimpleExtension.loadDefaults());
+  }
+
+  public ProtoExtendedExpressionConverter(SimpleExtension.ExtensionCollection extensionCollection) {
+    this.extensionCollection = extensionCollection;
+  }
+
+  private final ProtoTypeConverter protoTypeConverter =
+      new ProtoTypeConverter(
+          new ExtensionCollector(), ImmutableSimpleExtension.ExtensionCollection.builder().build());
+
+  public ExtendedExpression from(io.substrait.proto.ExtendedExpression extendedExpression) {
+    // fill in simple extension information through a discovery in the current proto-extended
+    // expression
+    ExtensionLookup functionLookup =
+        ImmutableExtensionLookup.builder().from(extendedExpression).build();
+
+    NamedStruct baseSchemaProto = extendedExpression.getBaseSchema();
+
+    io.substrait.type.NamedStruct namedStruct =
+        io.substrait.type.NamedStruct.fromProto(baseSchemaProto, protoTypeConverter);
+
+    ProtoExpressionConverter protoExpressionConverter =
+        new ProtoExpressionConverter(
+            functionLookup, this.extensionCollection, namedStruct.struct(), null);
+
+    List<ExtendedExpression.ExpressionReference> expressionReferences = new ArrayList<>();
+    for (ExpressionReference expressionReference : extendedExpression.getReferredExprList()) {
+      if (expressionReference.getExprTypeCase().getNumber() == 1) { // Expression
```

Comment:
optional: is there another provided option that's better than comparing with `getNumber()`? If not, then I think this is okay.

Reply:
IMO the nicer way of doing this is with a switch statement like
```java
      switch (expressionReference.getExprTypeCase()) {
        case EXPRESSION -> {
         ...
        }
        case MEASURE -> {
        ...
        }
        default -> throw new UnsupportedOperationException(
            "Only Expression or Aggregate Function type are supported in conversion from proto Extended Expressions");
      }
```

which avoids the magic number problem and is easier to read than if-else-if (at least for me).

Reply:
+1